lib/src/block_syntaxes/fenced_code_block_syntax.dart

```diff
@@ -19,77 +18,104 @@ class FencedCodeBlockSyntax extends BlockSyntax {
-    final infoString = match.group(2);
-    // From the CommonMark spec:
-    //
-    // > If the info string comes after a backtick fence, it may not contain
-    // > any backtick characters.
-    return codeFence.codeUnitAt(0) != $backquote ||
-        !infoString!.codeUnits.contains($backquote);
+  Node parse(BlockParser parser) {
+    final openingFence = _FenceMatch.fromMatch(pattern.firstMatch(
+      escapePunctuation(parser.current),
+    )!);
+
+    var text = parseChildLines(
+      parser,
+      openingFence.marker,
+      openingFence.indent,
+    ).join('\n');
+
+    if (parser.document.encodeHtml) {
+      text = escapeHtml(text);
+    }
+    if (text.isNotEmpty) {
+      text = '$text\n';
+    }
+
+    final code = Element.text('code', text);
+    if (openingFence.hasLanguage) {
+      var language = decodeHtmlCharacters(openingFence.language);
+      if (parser.document.encodeHtml) {
+        language = escapeHtmlAttribute(language);
+      }
+      code.attributes['class'] = 'language-$language';
+    }
+
+    return Element('pre', [code]);
   }
 
-  @override
-  List<String> parseChildLines(BlockParser parser, [String? endBlock]) {
-    endBlock ??= '';
+  String _removeIndentation(String content, int length) {
+    final text = content.replaceFirst(RegExp('^\\s{0,$length}'), '');
+    return content.substring(content.length - text.length);
+  }
 
+  @override
+  List<String> parseChildLines(
+    BlockParser parser, [
+    String openingMarker = '',
+    int indent = 0,
+  ]) {
     final childLines = <String>[];
+
     parser.advance();
 
+    _FenceMatch? closingFence;
     while (!parser.isDone) {
       final match = pattern.firstMatch(parser.current);
-      if (match == null || !match[1]!.startsWith(endBlock)) {
-        childLines.add(parser.current);
+      closingFence = match == null ? null : _FenceMatch.fromMatch(match);
+
+      // Closing code fences cannot have info strings:
+      // https://spec.commonmark.org/0.30/#example-147
+      if (closingFence == null ||
+          !closingFence.marker.startsWith(openingMarker) ||
+          closingFence.hasInfo) {
+        childLines.add(_removeIndentation(parser.current, indent));
         parser.advance();
       } else {
         parser.advance();
         break;
       }
     }
 
+    // https://spec.commonmark.org/0.30/#example-127
+    // https://spec.commonmark.org/0.30/#example-128
+    if (closingFence == null && childLines.last.trim().isEmpty) {
+      childLines.removeLast();
+    }
+
     return childLines;
   }
+}
 
-  @override
-  Node parse(BlockParser parser) {
-    // Get the syntax identifier, if there is one.
-    final match = pattern.firstMatch(parser.current)!;
-    final endBlock = match.group(1);
-    var infoString = match.group(2)!;
+class _FenceMatch {
+  _FenceMatch.fromMatch(RegExpMatch match) : indent = match[1]!.length {
+    if (match.namedGroup('backtick') != null) {
+      marker = match.namedGroup('backtick')!;
```

Comment:
Rather than initialize these `late final` fields in the constructor body, let's go ahead and make this a factory constructor, and add a `_FencedMatch._` constructor that assigns `marker` and `info` which can be `final` fields.

Reply:
Thanks! really cool tricks.